### ISSUE TITLE
As a machine learning engineer, I want to do research on choosing good models for replay analysis, so that I can make accurate and relevant predictions.

### ISSUE THREAD

**Issue by chrishollandaise** [stratosphererl/stratosphere#18]
# Acceptance Criteria
- [x] Given a set of replay data and a problem is well understood
- [x] When I theorize what models are the best fit for the data
- [x] Then I can begin to test my hypotheses in an educated way

# Estimation of Work
- 3 points
- Matt: 70 mins over 1 days --> Plan: Oct. 24th, 2022 (70 mins)

# Tasks
- [x] Research
- [x] Discuss

# Notes
- For each model, understand what kind of problem it is (e.g. clustering, optimization, recommendation).
- What sort of model best fits the data (e.g. supervised, unsupervised methods?

**Comment by chrishollandaise** [comment]
# Chris' Research
## Principal Component Analysis (PCA)
### What is it?
Principle component analysis is a method of dimensionality reduction. An example of PCA in action would be shirt sizes. Features of a shirt may be neck circumference, sleeve length, and body length, but clothing companies reduce these variables into shirt sizes (S, M, L).

### How is it useful?
This could be useful in reducing the dimensions of our replay to more critical features. This would make it easier to analyze the data and find patterns.

### Relevant resources?
[StatQuest](https://www.youtube.com/watch?v=FgakZw6K1QQ)
[skLearn](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html)

## Principal Component Analysis (PCA)
### What is it?
### How is it useful?
### Relevant resources?

**Comment by chrishollandaise** [comment]
# Chris' Research

## Principles of Unsupervised Learning
There are four common tasks that occur in unsupervised learning: reducing dimensions, detecting anomalies, clustering data, and association rule learning

### Dimensionality Reduction Techniques: 
- [Principal Component Analysis (PCA)](https://www.youtube.com/watch?v=FgakZw6K1QQ&t=167s)
- [Linear Discriminant Analysis (LDA)](https://www.youtube.com/watch?v=azXCzI57Yfc&t=455s)
- [Singular Value Decomposition (SVD)](https://gregorygundersen.com/blog/2018/12/10/svd/)
- [Manifold Learning ](https://www.youtube.com/watch?v=MnRskV3NY1k)
- [Autoencoders](https://www.youtube.com/watch?v=3jmcHZq3A5s)

<table>
    <thead>
        <tr>
          <td><b>Technique<b></td>
          <td><b>Pros<b></td>
          <td><b>Cons<b></td>
          <td><b>Additional Use Cases<b></td>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>Principal Component Analysis (PCA)</td>
            <td>
                <ul>
                    <li>Simple and easy to implement</li>
                    <li>Does not require any assumptions about the data</li>
                    <li>Can be used to reduce the dimensionality of data while still preserving the most important information</li>
                </ul>
            </td>
            <td>
                <ul>
                    <li>May not be able to capture all the information in the data if the data is not linearly separable</li>
                    <li>May not be able to reduce the dimensionality of the data as much as other methods</li>
                </ul>
            </td>
            <td>
                <ul>
                </ul>
            </td>
        </tr>
        <tr>
            <td>Linear Discriminant Analysis (LDA)</td>
            <td>
                <ul>
                    <li>Can be more effective at reducing the dimensionality of data than PCA</li>
                    <li>Can be used for classification tasks</li>
                </ul>
            </td>
            <td>
                <ul>
                    <li>Assumes that the data is linearly separable</li>
                    <li>May not be able to capture all the information in the data if the data is not linearly separable</li>
                </ul>
            </td>
            <td>
            </td>
        </tr>
        <tr>
            <td>Singular Value Decomposition (SVD)</td>
            <td>
                <ul>
                    <li>Can be used for dimensionality reduction, denoising, and image compression</li>
                </ul>
            </td>
            <td>
                <ul>
                    <li>May not be able to capture all the information in the data if the data is not linearly separable</li>
                </ul>
            </td>
            <td>
                <ul>
                    <li>Image compression</li>
                    <li>Denoising</li>
                </ul>
            </td>
        </tr>
        <tr>
            <td>Manifold Learning</td>
            <td>
                <ul>
                    <li>Can be used for nonlinear dimensionality reduction</li>
                    <li>Can be used for data visualization</li>
                </ul>
            </td>
            <td>
                <ul>
                    <li>May be difficult to implement</li>
                    <li>May not be able to capture all the information in the data</li>
                </ul>
            </td>
            <td>
                <ul>
                </ul>
            </td>
        </tr>
        <tr>
            <td>Autoencoders</td>
            <td>
                <ul>
                    <li>Can be used for nonlinear dimensionality reduction</li>
                    <li>Can be used for data denoising</li>
                    <li>Can be used for feature extraction</li>
                    <li>Can be used for anomaly detection</li>
                </ul>
            </td>
            <td>
                <ul>
                    <li>May be difficult to implement</li>
                    <li>May not be able to capture all the information in the data</li>
                </ul>
            </td>
            <td>
                <ul>
                    <li>Data denoising</li>
                    <li>Feature extraction</li>
                    <li>Anomaly detection</li>
                </ul>
            </td>
        </tr>
    </tbody>
</table>

### Anomaly Detection Algorithms:
- [Isolation Forest](https://www.youtube.com/watch?v=E0M73NTg3w0)
- [Local Outlier Factor (LOF)](https://www.youtube.com/watch?v=TTQ-9PWlU_w)


<table>
    <thead>
        <tr>
          <td><b>Technique<b></td>
          <td><b>Pros<b></td>
          <td><b>Cons<b></td>
          <td><b>Additional Use Cases<b></td>
    </thead>
    <body>
        <tr>
            <td>Isolation Forest</td>
            <td>
                <ul>
                    <li>Finding outliers in high dimensional data</li>
                </ul>
            </td>
            <td>
                <ul>
                    <li>Takes longer to train than LOF</li>
                    <li>May not perform as well as LOF on low dimensional data</li>
                </ul>
            </td>
            <td>
                <ul>
                    <li>Detecting fraudulent transactions</li>
                    <li>Detecting malicious network activity</li>
                </ul>
            </td>
        </tr>
        <tr>
            <td>Local Outlier Factor</td>
            <td>
                <ul>
                    <li>Runs faster than Isolation Forest</li>
                    <li>May perform better than Isolation Forest on low dimensional data</li>
                </ul>
            </td>
            <td>
                <ul>
                    <li>May not be as effective on high dimensional data</li>
                </ul>
            </td>
            <td>
                <ul>
                    <li>Detecting unusual patterns in data</li>
                    <li>Monitoring machine learning models for concept drift</li>
                </ul>
            </td>
        </tr>
    </body>
</table>

### Clustering Algorithms:
- [K-Means](https://www.youtube.com/watch?v=4b5d3muPQmA&t=1s)
- [Hierarchical Clustering](https://www.youtube.com/watch?v=7xHsRkOdVwo) and [Spectral Clustering](https://www.youtube.com/watch?v=uxsDKhZHDcc)
- [DBSCAN ](https://www.youtube.com/watch?v=RDZUdRSDOok)and [OPTICS](https://www.youtube.com/watch?v=9NyGB9jNf1E)
- [Affinity Propagation](https://towardsdatascience.com/understanding-affinity-propagation-clustering-and-implementation-with-python-c1e78fe57cde)
- [Mean Shift](https://www.youtube.com/watch?v=Evc53OaDTFc) and [BIRCH](https://www.youtube.com/watch?v=bHn0ECepk9Y)
- [Gaussian Mixture Models](https://www.youtube.com/watch?v=JNlEIEwe-Cg)

<table>
    <thead>
        <tr>
          <td><b>Technique<b></td>
          <td><b>Pros<b></td>
          <td><b>Cons<b></td>
          <td><b>Additional Use Cases<b></td>
    </thead>
    <body>
        <tr>
            <td>K-Means</td>
            <td>
                <ul>
                     <li>Simple, easy to understand and implement</li>
                     <li>Efficient in terms of computational cost (O(IKmn))</li>
                </ul>
            </td>
            <td>
                <ul>
                     <li>Difficult to predict K-value</li>
                     <li>Requires prior knowledge of number of clusters</li>
                     <li>sensative to outliers and scale due to its reliance on Euclidean distance</li>
                </ul>
            </td>
            <td>
                <ul>
                     <li>Market Segmentation</li>
                     <li>Speech Recognition</li>
                     <li>Image Compression</li>
                </ul>
            </td>
        </tr>
        <tr>
            <td>Hierarchical Clustering</td>
            <td>
                <ul>
                     <li>No pre-specified number of clusters</li>
                     <li>Clusters can be visualized with a dendrogram</li>
                </ul>
            </td>
            <td>
                <ul>
                     <li>Computationally expensive</li>
                     <li>Algorithm cannot be reversed</li>
                </ul>
            </td>
            <td>
                <ul>
                     <li>Document Clustering</li>
                     <li>Image Segmentation</li>
                </ul>
            </td>
        </tr>
        <tr>
            <td>Spectral Clustering</td>
            <td>
                <ul>
                     <li>Works well with higher dimensional data</li>
                     <li>Can find non-convex clusters</li>
                     <li>Cluster assignments are more robust to noise</li>
                </ul>
            </td>
            <td>
                <ul>
                     <li>Computationally expensive</li>
                     <li>Difficult to tune the algorithm</li>
                </ul>
            </td>
            <td>
                <ul>
                     <li>Image Segmentation</li>
                     <li>Document Clustering</li>
                     <li>Bioinformatics</li>
                </ul>
            </td>
        </tr>
        <tr>
            <td>DBSCAN</td>
            <td>
                <ul>
                     <li>Can find arbitrarily shaped clusters</li>
                     <li>Clusters of different densities can be found</li>
                     <li>No need to specify the number of clusters</li>
                </ul>
            </td>
            <td>
                <ul>
                     <li>Computationally expensive with large datasets</li>
                     <li>Difficult to choose appropriate values for epsilon and min_samples parameters</li>
                </ul>
            </td>
            <td>
                <ul>
                     <li>Anomaly Detection</li>
                     <li>Outlier Detection</li>
                </ul>
            </td>
        </tr>
        <tr>
            <td>OPTICS</td>
            <td>
                <ul>
                     <li>Can find arbitrarily shaped clusters</li>
                     <li>Does not require a predefined number of clusters</li>
                     <li>Clusters of different densities can be found</li>
                </ul>
            </td>
            <td>
                <ul>
                     <li>Computationally expensive with large datasets</li>
                     <li>Difficult to choose appropriate values for epsilon and min_samples parameters</li>
                </ul>
            </td>
            <td>
                <ul>
                     <li>Anomaly Detection</li>
                     <li>Outlier Detection</li>
                </ul>
            </td>
        </tr>
        <tr>
            <td>Affinity Propagation</td>
            <td>
                <ul>
                     <li>No need to specify the number of clusters</li>
                     <li>Clusters can be visualized with a dendrogram</li>
                </ul>
            </td>
            <td>
                <ul>
                     <li>Computationally expensive</li>
                     <li>Difficult to choose appropriate values for preference parameter</li>
                </ul>
            </td>
            <td>
                <ul>
                     <li>Document Clustering</li>
                     <li>Image Segmentation</li>
                </ul>
            </td>
        </tr>
        <tr>
            <td>Mean Shift</td>
            <td>
                <ul>
                     <li>No need to specify the number of clusters</li>
                     <li>Clusters can be visualized with a dendrogram</li>
                </ul>
            </td>
            <td>
                <ul>
                     <li>Computationally expensive</li>
                     <li>Difficult to choose appropriate values for bandwidth parameter</li>
                </ul>
            </td>
            <td>
                <ul>
                     <li>Image Segmentation</li>
                     <li>Object Tracking</li>
                </ul>
            </td>
        </tr>
        <tr>
            <td>BIRCH</td>
            <td>
                <ul>
                     <li>Scales linearly with the number of data points</li>
                     <li>No need to specify the number of clusters</li>
                </ul>
            </td>
            <td>
                <ul>
                     <li>Difficult to choose appropriate values for threshold parameter</li>
                </ul>
            </td>
            <td>
                <ul>
                     <li>Document Clustering</li>
                     <li>Pattern Recognition</li>
                </ul>
            </td>
        </tr>
        <tr>
            <td>Gaussian Mixture Models</td>
            <td>
                <ul>
                     <li>Can find arbitrarily shaped clusters</li>
                     <li>Clusters can overlap</li>
                </ul>
            </td>
            <td>
                <ul>
                     <li>Computationally expensive</li>
                     <li>Difficult to choose appropriate values for variance parameters</li>
                </ul>
            </td>
            <td>
                <ul>
                     <li>Speech Recognition</li>
                     <li>Image Segmentation</li>
                </ul>
            </td>
        </tr>
    </body>
</table>

### Association Rule Learning Algorithms:
> Machine learning algorithms that are used to discover relationships between variables in a dataset. These algorithms are used to find patterns in data and can be used for a variety of tasks such as classification, prediction, and clustering.

- [Apriori](https://www.youtube.com/watch?v=WGlMlS_Yydk)
- [Equivalence Class Transformation (Eclat)](https://www.youtube.com/watch?v=oBiq8cMkTCU)
- [FP-Growth](https://www.youtube.com/watch?v=GcgfSJAaBto)

<table>
    <thead>
        <tr>
          <td><b>Technique<b></td>
          <td><b>Pros<b></td>
          <td><b>Cons<b></td>
          <td><b>Additional Use Cases<b></td>
    </thead>
    <body>
        <tr>
            <td>Apriori</td>
            <td>
                <ul>
                     <li>Simple to understand and implement</li> 
                     <li>Can be used to find frequent itemsets in a dataset</li>
                     <li>Can be used to generate association rules</li>
                </ul>
            </td>
            <td>
                <ul>
                    <li>Scales linearly with the number of items in the dataset</li>
                    <li>May not be able to find all frequent itemsets in some datasets</li>
                </ul>
            </td>
            <td>
                <ul>
                    <li>Market basket analysis</li>
                    <li>Fraud detection</li>
                </ul>
            </td>
        </tr>
        <tr>
            <td>Eclat</td>
            <td>
                <ul>
                    <li>Simple to understand and implement</li>
                    <li>Can be used to find frequent itemsets in a dataset</li>
                    <li>Can be used to generate association rules</li>
                </ul>
            </td>
            <td>
                <ul>
                    <li>Scales linearly with the number of items in the dataset</li>
                    <li>May not be able to find all frequent itemsets in some datasets</li>
                </ul>
            </td>
            <td>
                <ul>
                    <li>Market basket analysis</li>
                    <li>Fraud detection</li>
                </ul>
            </td>
        </tr>
        <tr>
            <td>FP-Growth</td>
            <td>
                <ul>
                    <li>Can be used to find frequent itemsets in a dataset</li>
                    <li>Can be used to generate association rules</li>
                </ul>
            </td>
            <td>
                <ul>
                    <li>Scales linearly with the number of items in the dataset</li>
                    <li>May not be able to find all frequent itemsets in some datasets</li>
                </ul>
            </td>
            <td>
                <ul>
                    <li>Market basket analysis</li>
                    <li>Fraud detection</li>
                </ul>
            </td>
        </tr>
    </body>
</table>

## Discussion
I believe that we can actually use both supervised and unsupervised learning. It just depends on the task. For example, if we want to predict how likely actors in a replay are to score a goal given their position, we can look at games where players have won the match and see their position. We have labels for the players like MMR and such, so we can also use supervised learning. 

**Supervised models:** random forests, decision trees, support vector machines, linear and logistic regression, gradient boosting, naive bayes, linear discriminant analysis, neural networks

**Comment by matt89201824** [comment]
For playstyles / traits Machine Learning:
1. Create some basic traits which can be "manually" identified (i.e. enforcer == >20 car bumps, goalscorer = >2 goals scored, etc.)
2. Cluster players into groups using unsupervised learning
3. Manually identify a name for each group
4. If desired, label the player data within the core of the cluster as the name for the group chosen (do keep a copy with the data still unlabeled however)
5. Keep players outside of the core of any cluster as non-labeled (the model may, as replays are uploaded, begin a new cluster!)
6. We now have a dataset which we can use K-nearest neighbors (a supervised model)

We may want to have two datasets and models maintained here:
1. The unsupervised model with non-labeled data which simply looks to cluster players
2. And then a supervised model with the same data as the unsupervised model but with manually selected cluster names as labels
3. Given input data, the unsupervised model may adjust the number of clusters and the boundaries of clusters
4. This would then affect the data of the supervised model, affecting the output of that model
5. This mechanism may need manual tweaks every so often if new playstyles and strategies begin to be played in RL (hopefully because of our work!)

Flow:
Automatic data collection from uploaded replays > insertion of data into unsupervised model > unsupervised model updates, potentially modifying number of clusters and boundaries of clusters > this updates the labels on the data for supervised model > inputted data runs through supervised model > output provided

**Comment by cchurness** [comment]
Since Chris developed such a detailed list already, I'm just gonna mention some additional notes that I was looking into:

1. Dimensionality Reduction Techniques could be used as a preprocessor before our data gets to the "real" model. We have a lot of data, and our network frames will consist of a lot of features; DRT could help us reduce that and make the training process better.
2. It seems most of our unsupervised learning methods are likely to be clustering: traits, playstyle, and formation.
3. Certain models require different amounts of information: key moments would need the whole replay, player formations would need just a frame, and play style may require clustering a full replay into smaller bits (I think Dimension Reduction would be useful in this case)
4. Another thought: play style, traits, and mechanics could be combined into a hierarchical cluster, where player actions cluster into mechanics, which cluster into traits, which cluster into play style.
5. Key moments could be supervised referring to goals, saves, clears, etc. Then we could see if it categorizes anything other than those things as additional key moments
6. Apriori down the line could be used to suggest replays that the viewer might be interested in.
7. Autoencoders sounded cool for Dimension Reduction, but manifold might be easier than working with neural nets.
